### ISSUE TITLE
Automatic cleaning of media cache

### ISSUE THREAD

**Issue by Ekdohibs** [minetest/minetest#10598]
Currently, the media cache is not automatically cleaned (I have folders in my media folder that were created in 2014!). However, with the not-so-recent introduction of translation files as media, and probably the future introduction of server-sent lua files, a lot of media files will get useless relatively fast.

Therefore, I propose to introduce an automatic cleaning of the media cache, which would remove old files. There needs to be some thinking about the heuristics used, because we don't want to evict some useful media when connecting to servers with lots of custom media. If there are time-based heuristics, the time scale used should probably depend on media type: code and translations will probably have far shorter lifespans than textures or models.

**Comment by SmallJoker** [comment]
The perhaps easiest approach would be to delete files that were last accessed 6 months ago.

**Comment by red-001** [comment]
Last access timestamp might not be reliable, sometimes it's disabled for performance reasons.

**Comment by hecktest** [comment]
Move it all to one sqlite database, store the time and hash in there. At least you won't be trashing the filesystem with lots of tiny files.

**Comment by red-001** [comment]
Enjoy making sure it doesn't lock up when you have more than one client.

**Comment by hecktest** [comment]
Right. What about prepending a simple numerical timestamp to the file name, in the form of a few hex digits?

**Comment by sorcerykid** [comment]
Perhaps I'm missing something, but isn't this a client-side media cache? So why would database semaphores come into play? The clearing of media can be performed by the main thread when the application launches. And the updating of file hashes and timestamps can occur during the media download phase, which is already executed during a single thread.

Anyway, I really like this idea and would love to see it added to the engine at some point.

**Comment by Desour** [comment]
>So why would database semaphores come into play?

You can start multiple instances of the minetest client. All of them use the same media cache.

However, loading media files doesn't happen often, so one could just lock the whole cache, right?

Are the access timestamps reliable if we touch the media files every time?

If someone didn't use minetest for some months, it shouldn't delete nearly all cached files imo. The user can not join all their favourite games with different media at once. Maybe one should instead count the days in which minetest was started, and use this counter instead of a timestamp.

**Comment by erlehmann** [comment]
> Last access timestamp might not be reliable, sometimes it's disabled for performance reasons.

I think `atime` is reliable *enough*.

Rationale: Even if `atime` updates are disabled using the `relatime` mount option (which I think you are), Linux is hardcoded to still update the `atime` value, if the previous value is more than 24 hours in the past. This is good enough for simple, but effective cache eviction algorithms like:

* If the cache directory size is over a threshold, delete files – starting with files that have not been used the longest.
* Delete files that were not used for a long time exceeding a threshold.

**Comment by erlehmann** [comment]
One likely reason why this has not been addressed in a long time is that it is literally a one-liner on Linux:

`tmpreaper 30d cache` will remove times that have not been accessed in 30 days.

http://manpages.ubuntu.com/manpages/bionic/man8/tmpreaper.8.html
